From 9b921b3b822ebf63cf7d32297609b976e1d78d01 Mon Sep 17 00:00:00 2001
From: Parikshit Gune <quic_pgune@quicinc.com>
Date: Thu, 17 Oct 2024 14:29:15 +0530
Subject: [PATCH] QSDK: wifi: mac80211: Add init netdev ops for an
 interface.

Adding new ops for netdev init and netdev deinit in ieee80211_ops.

	1. init_interface ops to be called at interface ndo_init time
	2. deinit_interface ops to be called at interface ndo_deinit time
	3. This gives flexibility for vendors to link the initializations /
	   deinitialization of vender / driver specific objects at the time
	   of netdev init - deinit stages.

ndo_deinit is already present int mac80211 to process at the time of
interface deinit. Here, adding callback for ndo_init as well for
an interface.

Patch-dependency: None
Patch-work: None

Signed-off-by: Parikshit Gune <quic_pgune@quicinc.com>
Signed-off-by: Hari Chandrakanthan <quic_haric@quicinc.com>
---
 include/net/mac80211.h    | 10 ++++++++++
 net/mac80211/driver-ops.h | 24 ++++++++++++++++++++++++
 net/mac80211/iface.c      | 15 ++++++++++++++-
 3 files changed, 48 insertions(+), 1 deletion(-)

diff --git a/include/net/mac80211.h b/include/net/mac80211.h
index fe15526..a4a7fcf 100644
--- a/include/net/mac80211.h
+++ b/include/net/mac80211.h
@@ -4451,6 +4451,10 @@ struct ieee80211_ppe_vp_ds_params {
  *	Note that the driver must not assume that the vif driver_data is valid
  *	at this point, since the callback can be called during netdev teardown.
  * @change_mtu: Ops to change the interface MTU.
+ * @init_interface: Called at the time of netdev init stage. This gives drivers
+ *	a flexibility to init driver specific data at interface init time.
+ * @deinit_interface: Called at the time of netdev deinit stage. This gives drivers
+ *	a flexibility to deinit driver specific data at interface deinit time.
  */
 struct ieee80211_ops {
 	void (*tx)(struct ieee80211_hw *hw,
@@ -4835,6 +4839,12 @@ struct ieee80211_ops {
 	int (*change_mtu)(struct ieee80211_hw *hw,
 			  struct ieee80211_vif *vif,
 			  int mtu);
+	int (*init_interface)(struct ieee80211_hw *hw,
+			      struct ieee80211_vif *vif,
+			      struct net_device *dev);
+	int (*deinit_interface)(struct ieee80211_hw *hw,
+				struct ieee80211_vif *vif,
+				struct net_device *dev);
 };
 
 /**
diff --git a/net/mac80211/driver-ops.h b/net/mac80211/driver-ops.h
index 5073f19..4163e84 100644
--- a/net/mac80211/driver-ops.h
+++ b/net/mac80211/driver-ops.h
@@ -1562,6 +1562,30 @@ static inline int drv_net_setup_tc(struct ieee80211_local *local,
 	return ret;
 }
 
+static inline int drv_init_interface(struct ieee80211_local *local,
+				     struct ieee80211_sub_if_data *sdata,
+				     struct net_device *dev)
+{
+	int ret = 0;
+
+	if (local->ops->init_interface)
+		ret = local->ops->init_interface(&local->hw, &sdata->vif, dev);
+
+	return ret;
+}
+
+static inline int drv_deinit_interface(struct ieee80211_local *local,
+				       struct ieee80211_sub_if_data *sdata,
+				       struct net_device *dev)
+{
+	int ret = 0;
+
+	if (local->ops->deinit_interface)
+		ret = local->ops->deinit_interface(&local->hw, &sdata->vif, dev);
+
+	return ret;
+}
+
 int drv_change_vif_links(struct ieee80211_local *local,
 			 struct ieee80211_sub_if_data *sdata,
 			 u16 old_links, u16 new_links,
diff --git a/net/mac80211/iface.c b/net/mac80211/iface.c
index 68af17a..f3e0b15 100644
--- a/net/mac80211/iface.c
+++ b/net/mac80211/iface.c
@@ -924,9 +924,20 @@ static void ieee80211_teardown_sdata(struct ieee80211_sub_if_data *sdata)
 	ieee80211_link_stop(&sdata->deflink);
 }
 
+static int ieee80211_init(struct net_device *dev)
+{
+	struct ieee80211_sub_if_data *sdata = IEEE80211_DEV_TO_SUB_IF(dev);
+	struct ieee80211_local *local = sdata->local;
+
+	return drv_init_interface(local, sdata, dev);
+}
 static void ieee80211_uninit(struct net_device *dev)
 {
-	ieee80211_teardown_sdata(IEEE80211_DEV_TO_SUB_IF(dev));
+	struct ieee80211_sub_if_data *sdata = IEEE80211_DEV_TO_SUB_IF(dev);
+	struct ieee80211_local *local = sdata->local;
+
+	drv_deinit_interface(local, sdata, dev);
+	ieee80211_teardown_sdata(sdata);
 }
 
 static void
@@ -970,6 +981,7 @@ static const struct net_device_ops ieee80211_dataif_ops = {
 	.ndo_get_stats64	= ieee80211_get_stats64,
 	.ndo_setup_tc		= ieee80211_netdev_setup_tc,
 	.ndo_change_mtu		= ieee80211_change_mtu,
+	.ndo_init               = ieee80211_init,
 };
 
 static u16 ieee80211_monitor_select_queue(struct net_device *dev,
@@ -1021,6 +1033,7 @@ static const struct net_device_ops ieee80211_dataif_8023_ops = {
 	.ndo_get_stats64	= ieee80211_get_stats64,
 	.ndo_setup_tc		= ieee80211_netdev_setup_tc,
 	.ndo_change_mtu		= ieee80211_change_mtu,
+	.ndo_init               = ieee80211_init,
 };
 
 static bool ieee80211_iftype_supports_hdr_offload(enum nl80211_iftype iftype)
-- 
2.34.1

